### PR TITLE
Support edma3 for s32z270dc2 board

### PR DESCRIPTION
**Pull request by haduongquang**
This PR introduces NXP S32 EDMA3 driver for SoC NXP S32Z27 and enables its usage for board s32z270dc2

Tests:

tests\drivers\dma\chan_blen_transfer:
`SUITE PASS - 100.00% [dma_m2m]: pass = 4, fail = 0, skip = 0, total = 4 duration = 8.063 seconds`

tests\drivers\dma\chan_link_transfer:
`SUITE PASS - 100.00% [dma_m2m_link]: pass = 3, fail = 0, skip = 0, total = 3 duration = 6.047 seconds`

tests\drivers\dma\loop_transfer:
`SUITE PASS - 100.00% [dma_m2m_loop]: pass = 3, fail = 0, skip = 0, total = 3 duration = 1.103 seconds`

tests\drivers\dma\scatter_gather:
`SUITE PASS - 100.00% [dma_m2m_sg]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.076 seconds`

### CODE REVIEW

File: boards/nxp/s32z2xxdc2/doc/index.rst [hunk 1]
```diff
@@ -163,6 +165,16 @@ ADC is provided through ADC SAR controller with 2 instances. Each ADC SAR instan
 .. note::
    All channels of an instance only run on 1 group channel at the same time.
 
+EDMA
+====
+
+The EDMA modules are categorized into one EDMA4 instance (instance 3) and four EDMA3 instances
```
Comment (Dat-NguyenDuy) [inline boards/nxp/s32z2xxdc2/doc/index.rst:171]: The PR description said this supports eDMA3, why you mention eDMA4 here

File: tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	soc {
+		sram1: memory@35780000 {
+			compatible = "mmio-sram";
+			reg = <0x31780000 DT_SIZE_K(960)>;
+		};
+
+		sram2: memory@35870000 {
+			compatible = "zephyr,memory-region", "mmio-sram";
+			reg = <0x35870000 DT_SIZE_K(64)>;
+			zephyr,memory-region = "SRAM3";
```
Comment (Dat-NguyenDuy) [inline tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay:19]: `SRAM2`? or perhaps use `sram_nocache` for node lable and `SRAM_NOCACHE` for `zephyr,memory-region`. Use the same approach for all places

File: tests/drivers/dma/loop_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	soc {
+		sram1: memory@35780000 {
+			compatible = "mmio-sram";
+			reg = <0x31780000 DT_SIZE_K(960)>;
+		};
+
+		sram2: memory@35870000 {
+			compatible = "zephyr,memory-region", "mmio-sram";
+			reg = <0x35870000 DT_SIZE_K(64)>;
+			zephyr,memory-region = "SRAM3";
```
Comment (Dat-NguyenDuy) [inline tests/drivers/dma/loop_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay:19]: `SRAM2`?

File: tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/dt-bindings/memory-attr/memory-attr-arm.h>
+
+/ {
+	soc {
+		sram1: memory@35780000 {
```
Comment (Dat-NguyenDuy) [inline tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay:11]: You don't need to re-define the whole node here, just use something like this (but out of devicetree root `/{ }` )
```
&sram1 {
     reg = <0x31780000 DT_SIZE_K(960)>;
}
```
Do this for all places. Then please mention in the commit message that the non-cacheable memory used for DMA tests is split from the system SRAM

**Review by Dat-NguyenDuy — CHANGES_REQUESTED** [review]

File: boards/nxp/s32z2xxdc2/doc/index.rst (hunk 1 above)
Comment (haduongquang) [reply boards/nxp/s32z2xxdc2/doc/index.rst]: I removed eDMA4

File: tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay (hunk 2 above)
Comment (haduongquang) [reply tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay:14]: I updated

File: tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay (hunk 4 above)
Comment (haduongquang) [reply tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay]: Thanks, I updated.

File: tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay (hunk 2 above)
Comment (Dat-NguyenDuy) [reply tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay:14]: why not `sram2`?

File: tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay (hunk 4 above)
Comment (Dat-NguyenDuy) [reply tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay]: Where is this update `Then please mention in the commit message that the non-cacheable memory used for DMA tests is split from the system SRAM`? put something like that in commit msg or in code pls

Comment (haduongquang) [reply tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay]: I updated the commit msg also.

File: tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay (hunk 2 above)
Comment (haduongquang) [reply tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay]: I updated `sram_nocache` for node lable and `SRAMNOCACHE` for `zephyr,memory-region`.

**Comment by haduongquang** [comment]
Added commit to Split `tcdpool` per device that can be used for multiple instances.

**Review by Dat-NguyenDuy — DISMISSED** [review]

File: drivers/dma/dma_mcux_edma.c [hunk 5]
```diff
@@ -716,17 +714,26 @@ static int dma_mcux_edma_init(const struct device *dev)
 #define DMA_INIT(n)								\
 	DMAMUX_BASE_INIT_DEFINE(n)						\
 	static void dma_imx_config_func_##n(const struct device *dev);		\
+	static __aligned(32) EDMA_TCDPOOL_CACHE_ATTR edma_tcd_t			\
+	tcdpool##n[DT_INST_PROP(n, dma_channels)][CONFIG_DMA_TCD_QUEUE_SIZE];	\
```
Comment (manuargue) [inline drivers/dma/dma_mcux_edma.c:718]:
```suggestion
	dma_tcdpool_##n[DT_INST_PROP(n, dma_channels)][CONFIG_DMA_TCD_QUEUE_SIZE];	\
```

File: drivers/dma/dma_mcux_edma.c [hunk 6]
```diff
@@ -716,17 +714,26 @@ static int dma_mcux_edma_init(const struct device *dev)
...
 		CHANNELS_PER_MUX(n)						\
 		.irq_config_func = dma_imx_config_func_##n,			\
 		.dmamux_reg_offset = DT_INST_PROP(n, dmamux_reg_offset),	\
 		DMA_MCUX_EDMA_CHANNEL_GAP(n)					\
+		.tcdpool = tcdpool##n,						\
 	};									\
 										\
-	struct dma_mcux_edma_data dma_data_##n;					\
+	static struct call_back data_callback_##n[DT_INST_PROP(n, dma_channels)];\
+	static ATOMIC_DEFINE(channels_atomic_##n, DT_INST_PROP(n, dma_channels));\
```
Comment (manuargue) [inline drivers/dma/dma_mcux_edma.c:732]:
```suggestion
	static ATOMIC_DEFINE(dma_channels_atomic_##n, DT_INST_PROP(n, dma_channels));\
```

File: drivers/dma/dma_mcux_edma.c [hunk 7]
```diff
@@ -716,17 +714,26 @@ static int dma_mcux_edma_init(const struct device *dev)
...
 		.dma_channels = DT_INST_PROP(n, dma_channels),			\
 		CHANNELS_PER_MUX(n)						\
 		.irq_config_func = dma_imx_config_func_##n,			\
 		.dmamux_reg_offset = DT_INST_PROP(n, dmamux_reg_offset),	\
 		DMA_MCUX_EDMA_CHANNEL_GAP(n)					\
+		.tcdpool = tcdpool##n,						\
 	};									\
 										\
-	struct dma_mcux_edma_data dma_data_##n;					\
+	static struct call_back data_callback_##n[DT_INST_PROP(n, dma_channels)];\
```
Comment (manuargue) [inline drivers/dma/dma_mcux_edma.c:731]:
```suggestion
	static struct call_back dma_data_callback_##n[DT_INST_PROP(n, dma_channels)];\
```

**Review by manuargue — CHANGES_REQUESTED** [review]

File: drivers/dma/dma_mcux_edma.c (hunk 5 above)
Comment (haduongquang) [reply drivers/dma/dma_mcux_edma.c]: I updated

File: drivers/dma/dma_mcux_edma.c (hunk 6 above)
Comment (haduongquang) [reply drivers/dma/dma_mcux_edma.c]: I updated

File: drivers/dma/dma_mcux_edma.c (hunk 7 above)
Comment (haduongquang) [reply drivers/dma/dma_mcux_edma.c]: I updated

**Comment by haduongquang** [comment]
@teburd @manuargue @Dat-NguyenDuy please help me take a look into this PR. All comments addressed. Thanks

**Review by manuargue — DISMISSED** [review]

**Review by Dat-NguyenDuy — DISMISSED** [review]

**Comment by haduongquang** [comment]
Hi @teburd, Could you please help me take a look into this PR?

**Comment by haduongquang** [comment]
I rebased and fixed the conflict.

**Review by manuargue — APPROVED** [review]

**Review by Dat-NguyenDuy — DISMISSED** [review]

**Review by dleach02 — DISMISSED** [review]

**Comment by dleach02** [comment]
@teburd I've add @manuargue as a co-assignee so we can move forward with this PR.

**Comment by manuargue** [comment]
@haduongquang HAL pr merged, please rebase

**Review by manuargue — CHANGES_REQUESTED** [review]
blocking until https://github.com/zephyrproject-rtos/zephyr/pull/78314 is merged and update the manifest

**Comment by haduongquang** [comment]
> blocking until #78314 is merged and update the manifest

I rebased and updated the manifest

**Review by manuargue — APPROVED** [review]
dismiss https://github.com/zephyrproject-rtos/zephyr/pull/78368#pullrequestreview-2376856145, my mistake, first we need to merge this pr

**Review by Dat-NguyenDuy — APPROVED** [review]

**Pull request merged**